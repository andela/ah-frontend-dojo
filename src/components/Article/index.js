/* eslint-disable react/no-unused-state */
// react libraries
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// styles
import 'components/Loader/Loader.scss';
import 'components/Footer/Footer.scss';
import './Article.scss';
import './ArticleOnly.scss';

// components
import Navbar from 'components/NavBar';
import Footer from 'components/Footer';
// import ProgressLoader from 'components/Loader/ProgressLoader';

import { newArticle } from '../../store/actions/articleActions';


export class Article extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: '',
      body: '',
      tagList: [],
    };
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { title, description, body } = this.state;
    const articleData = {
      article: {
        title,
        description,
        body,
        tagList: [],
      },
    };

    const { createArticle } = this.props;

    createArticle(articleData);
    window.location = '/articles/';
  }

  render() {
    const { title, description, body } = this.state;
    return (
      <div className="canvas">
        <div className="landing">
          <Navbar />
          <div className="breadcrumbs" align="left">
            <ul>
              <li>
                <img src="http://localhost:8080/src/assets/images/breadcrumbs.png" alt="breadcrumbs" />
                /Create Article/
              </li>
            </ul>
          </div>
          {/* <ProgressLoader /> */}
          <div className="inputPanel" align="center">
            <div className="inputForm">
              <form onSubmit={this.handleSubmit}>
                <input type="text" name="title" placeholder="Title" value={title} onChange={this.handleChange} required />
                <input type="text" id="description" name="description" placeholder="Description" value={description} onChange={this.handleChange} required />
                <textarea type="text" name="body" placeholder="Body" value={body} onChange={this.handleChange} required />
                <button type="submit" name="createArtile">Create Article</button>
              </form>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

Article.propTypes = {
  createArticle: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => {
  const { article } = state.articleReducer;
  return { article };
};

export const mapDispatchToProps = dispatch => ({
  createArticle(articleData) { dispatch(newArticle(articleData)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(Article);
