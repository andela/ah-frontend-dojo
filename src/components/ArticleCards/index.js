/* eslint-disable react/prop-types */
// react libraries
import React, { Component } from 'react';

// styles
import './ArticleCardLarge.scss';

// components
import Tagslist from 'components/TagsList';
import { defaultImgURL, formatDate, readMinutes } from 'utils';
import ArticleLikeDislike from 'components/ArticleLikeDislike';
import Comments from 'components/Comments';

export default class ArticleCardLarge extends Component {
  render() {
    const {
      title,
      body,
      description,
      author,
      updatedAt,
      time_to_read,
    } = this.props;
    const articleImg = `${defaultImgURL}/images/banner.jpg`;
    return (
      <div className="mainCanvas">
        <div className="cardCanvas">
          <div className="titlePanel">
            <p>{ title }</p>
          </div>
          <div className="imagePanel">
            <img src={articleImg} alt="article img" />
          </div>
          <div className="flexedRow">
            <div className="columnPanel aurthorPanel">
              <div className="authorAvartar">
                <img src="http://localhost:8080/src/assets/images/avartar.jpg" alt="article img" />
              </div>
              <div className="authorName">
                <p align="left">
                  <b>{ author }</b>
                  <br />
                  <button type="submit" className="buttonSmall buttonLarge">
                    View profile
                  </button>
                </p>
              </div>
            </div>
            <div className="columnPanel align-inText-right lightedParagraph">
              <p>
                <b>Date: </b>
                { formatDate(updatedAt) }
              </p>
              <p>
                { readMinutes(time_to_read) }
                Read
              </p>
            </div>
          </div>
          <div className="bodyPanel">
            <p className="description">{ description }</p>
            <p>{ body }</p>
          </div>
          <div className="additionalDetails inlinePosition">
            <p>
              <b>Average Rating: </b>
              3.9 (2000)
            </p>
            <p>
              <b>Views: </b>
              (0000)
              {/* { read_stats[0].views } */}
            </p>
          </div>
          {/* articleLikeDislike component here */}
          <ArticleLikeDislike />
          {/* tagsList component here */}
          <Tagslist />
          <div className="splitLine" />
          <div className="commentsPanel">
            <Comments />
          </div>
        </div>
      </div>
    );
  }
}
