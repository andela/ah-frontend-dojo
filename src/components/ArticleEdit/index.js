// react libraries
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// styles
import 'components/Loader/Loader.scss';
import 'components/Footer/Footer.scss';
import './ArticleEdit.scss';

// components
import Navbar from 'components/NavBar';
import Footer from 'components/Footer';
import { getOneArticle, editArticle } from 'store/actions/articleActions';
import ProgressLoader from 'components/Loader/ProgressLoader';


export class ArticleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: '',
      body: '',
      tagList: [],
    };
  }

  componentDidMount() {
    const { getArticle, match: { params: { id } }, article } = this.props;
    getArticle(id);

    this.setState(prevState => ({ ...prevState, article }));
  }


  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      title,
      description,
      body
    } = this.state;
    const articleData = {
      title,
      description,
      body,
      tagList: [],
    };
    const { updateArticle, article } = this.props;

    if (articleData.title === '') {
      articleData.title = article.title;
    }
    if (articleData.description === '') {
      articleData.description = article.description;
    }
    if (articleData.body === '') {
      articleData.body = article.body;
    }

    const articleObject = {
      id: article.id,
      article: articleData,
    };
    updateArticle(articleObject);
  }

  cancelArticleEdit = () => {
    const { article } = this.props;
    window.location = `/article/${article.id}`;
  }

  render() {
    const { article, editProgress } = this.props;
    return (
      <div className="canvas">
        <div className="landing">
          <Navbar />
          <div className="breadcrumbs" align="left">
            <ul>
              <li>
                <img
                  src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099269/breadcrumbs_qot92i.png"
                  alt="breadcrumbs"
                />
                /Edit Article/
              </li>
            </ul>
          </div>
          {editProgress && <ProgressLoader />}
          <div className="inputPanel" align="center">
            <div className="inputForm">
              <form onSubmit={this.handleSubmit}>
                <input type="text" name="title" defaultValue={article.title} onChange={this.handleChange} />
                <textarea
                  type="text"
                  name="description"
                  id="description"
                  onChange={this.handleChange}
                  placeholder={article.description}
                />
                <textarea
                  type="text"
                  name="body"
                  value={article.body}
                />
                <div className="editButtons">
                  <button
                    type="submit"
                    className=""
                    id="edit"
                    name="updateArticle"
                  >
                      Save Changes
                  </button>
                  <button type="button" className="" id="cancel" onClick={this.cancelArticleEdit} name="cancel">Cancel</button>
                </div>
              </form>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

ArticleEdit.propTypes = {
  getArticle: PropTypes.func.isRequired,
  updateArticle: PropTypes.func.isRequired,
  editProgress: PropTypes.bool,
  match: PropTypes.shape(
    {
      params: PropTypes.shape(
        {
          id: PropTypes.isRequired,
        }
      )
    }
  ).isRequired,
  article: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    description: PropTypes.string,
    body: PropTypes.string,
  })
};

ArticleEdit.defaultProps = {
  editProgress: false,
  article: {
    id: 61,
    title: '',
    description: '',
    body: '',
  }
};

export const mapStateToProps = (state) => {
  const { article, editProgress } = state.articleReducer;
  const articleId = state.articleReducer.article.id;
  return { article, articleId, editProgress };
};

export const mapDispatchToProps = dispatch => ({
  getArticle: (articleData) => { dispatch(getOneArticle(articleData)); },
  updateArticle: (articleData) => { dispatch(editArticle(articleData)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleEdit);
