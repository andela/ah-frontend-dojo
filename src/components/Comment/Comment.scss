@import 'assets/MainStyle.scss';

.comment_form {
  > form {
    width: 100%;
    //padding: none;
    background: none;

    & textarea {
      outline: none;
      width: 100%;
      border: 0.06rem #2D658623 solid;
      background: $white;
      border-radius: 0.7rem;
      padding: 0.5rem;
      height: 4rem;

      &:hover, &:focus {
        border: 0.06rem #2D658656 solid;
      }
    }

    & i {
      color: #C21F39;
    }
  }
}

.comment {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 1.2rem;
  //box-shadow: 0 0 0 0.1rem rgba(34, 34, 34, 0.06);
  border-radius: 5%;
  padding: 0.2rem;
  margin-bottom: 0.8rem;

  &__img {
    grid-column: 1/ span 1;
    grid-row: 1;
    justify-self: end;
    top: 50%;
    width: 4.375rem;
    box-shadow: 0 0 0.4rem rgba(0,0,0,.6);
    border-radius: 50%;
  }

  &__header {
    grid-column: 2 /span 3;
    grid-row: 1;
    display: inline-grid;
    grid-template-columns: repeat(3, 1fr);
    align-items: center;
    border-bottom: 1px solid #E5E5E5;

    &__author {
      grid-column: span 1;
      grid-row: 1;
      color: $primary;
      transition: color .3s ease-in-out;
      justify-self: start;

      &:hover {
        color: #222;
        text-decoration: underline;
      }

    }

    &__timestamp {
      grid-column: 2/span 2;
      grid-row: 1;
      color: #AEADAD;
      justify-self: end;
    }

  }

  &__body {
    grid-column: 2/span 3;
    grid-row: 2;
    display: block;
    color: rgba(34, 34, 34, 0.64);
  }

  &__footer {
    grid-column: 2/span 3;
    grid-row: 3;
    padding-bottom: 0.8rem;

    > .delete {
      color: #FF492E;
      border: none;

      &:hover {
        border: none;
        background-color: transparent;
      }

    }

    &__actions {
      justify-self: end;
    }
  }
}
