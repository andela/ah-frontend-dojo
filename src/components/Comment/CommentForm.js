import React, { Component } from 'react';

// third party libraries
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

// comment actions
import createComment from 'store/actions/commentActions';
import Button from 'components/Button';

export class CommentForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      commentBody: '',
      error: '',
    };
  }

  handleCommentSubmit = (e) => {
    e.preventDefault();


    const { commentBody } = this.state;
    const { createNewComment, articleSlug } = this.props;

    if (articleSlug && commentBody) {
      createNewComment(articleSlug, commentBody);
      this.setState({ commentBody: '' });
    }
  };

  handleChange = (e) => {
    const { name, value } = e.target;

    this.setState({ [name]: value });
  };

  renderCommentError = error => (
    error ? <div className="error-msg text-danger">{ error }</div> : ''
  );

  renderCommentSuccess = success => (success ? <div className="success-msg text-success">Comment added successfully</div> : '');


  render() {
    const { commentBody, error } = this.state;
    const { success, username } = this.props;

    return (
      <div className="comment_form">
        {this.renderCommentError(error)}
        {this.renderCommentSuccess(success)}

        <form onSubmit={this.handleCommentSubmit}>
          <div>
            <textarea
              placeholder="Comment"
              rows="3"
              required
              name="commentBody"
              value={commentBody}
              onChange={this.handleChange}
            />
          </div>
          <div>
            <Button inputType="submit" btnClass="btn" btnEvent={this.handleCommentSubmit} btnName="Post Comment" disabled={!username} />
            {!username && (
              <i>
                {' login to comment'}
              </i>
            )}
          </div>
        </form>
      </div>
    );
  }
}

// default props
CommentForm.defaultProps = {
  username: '',
  success: false,
};

// Props Validation
CommentForm.propTypes = {
  createNewComment: PropTypes.func.isRequired,
  articleSlug: PropTypes.string.isRequired,
  success: PropTypes.bool,
  username: PropTypes.string,
};

// mapState to props
export const mapStateToProps = (state) => {
  const { commentReducer: { success }, loginReducer: { user: { username } } } = state;
  return { success, username };
};

// map dispatch to props
export const mapDispatchToProps = dispatch => ({
  createNewComment(articleSlug, commentBody) {
    dispatch(createComment(articleSlug, commentBody));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(CommentForm);
