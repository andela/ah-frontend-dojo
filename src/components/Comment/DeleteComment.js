// React library
import React, { Fragment } from 'react';

// Third-party libraries
import PropTypes from 'prop-types';

// components
import Button from 'components/Button';

const DeleteComment = ({
  commentId, author, articleSlug, deleteComment, authenticatedUsername
}) => (
  <Fragment>
    {authenticatedUsername === author && (
    <Button
      btnClass="fa fa-trash btn delete"
      btnEvent={(event) => {
        event.preventDefault();
        deleteComment(articleSlug, commentId);
      }}
      btnName=""
      disabled={authenticatedUsername !== author}
    />
    )}
  </Fragment>
);

// PropType validation
DeleteComment.propTypes = {
  authenticatedUsername: PropTypes.string.isRequired,
  articleSlug: PropTypes.string.isRequired,
  commentId: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  deleteComment: PropTypes.func.isRequired,
};

export default DeleteComment;
