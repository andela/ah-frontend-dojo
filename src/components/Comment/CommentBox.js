// react
import React from 'react';
import PropTypes from 'prop-types';
import DeleteComment from 'components/Comment/DeleteComment';
import LikeDislikeCommentd from '../LikeDislikeComment';


const date = timestamp => (new Date(timestamp).toString().slice(0, 25));
const CommentBox = (
  {
    id, author: { username, image },
    created_at, body, articleSlug, deleteComment, authenticatedUsername, likeStatus, dislikeStatus
  }
) => (
  <div className="comment">
    <img className="comment__img" alt="author" src={image} />
    <div className="comment__header">
      <p className="comment__header__author">{username}</p>
      <span className="comment__header__timestamp">{date(created_at)}</span>
    </div>
    <div className="comment__body">{body}</div>
    <LikeDislikeCommentd
      likeStatus={likeStatus}
      dislikeStatus={dislikeStatus}
      articleSlug={articleSlug}
      commentId={id}
    />
    <div className="comment_footer">
      {' '}
      {}
    </div>
    <div className="comment__footer">
      <DeleteComment
        author={username}
        id={id}
        articleSlug={articleSlug}
        deleteComment={deleteComment}
        authenticatedUsername={authenticatedUsername}
      />
    </div>

  </div>
);

CommentBox.propTypes = {
  authenticatedUsername: PropTypes.string.isRequired,
  deleteComment: PropTypes.func.isRequired,
  articleSlug: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  body: PropTypes.string.isRequired,
  created_at: PropTypes.string.isRequired,
  author: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
  likeStatus: PropTypes.bool,
  dislikeStatus: PropTypes.bool,
};

CommentBox.defaultProps = {
  likeStatus: false,
  dislikeStatus: false,
};

export default CommentBox;
