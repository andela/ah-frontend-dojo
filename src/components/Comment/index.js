import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { readComments, removeComment } from 'store/actions/commentActions';

import CommentForm from 'components/Comment/CommentForm';
import CommentBox from 'components/Comment/CommentBox';

import './Comment.scss';

export class Comment extends Component {
  static propTypes = {
    username: PropTypes.string.isRequired,
    articleSlug: PropTypes.string.isRequired,
    commentsCount: PropTypes.number.isRequired,
    getComments: PropTypes.func.isRequired,
    deleteComment: PropTypes.func.isRequired,
    comments: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        created_at: PropTypes.string.isRequired,
        updated_at: PropTypes.string.isRequired,
        body: PropTypes.string.isRequired,
        author: PropTypes.shape({
          firstname: PropTypes.string.isRequired,
          lastname: PropTypes.string.isRequired,
          bio: PropTypes.string.isRequired,
          image: PropTypes.string.isRequired,
          email: PropTypes.string.isRequired,
        }).isRequired,
        highlighted_text: PropTypes.string,
        start_index: PropTypes.string,
        end_index: PropTypes.string,
        likeStatus: PropTypes.bool,
        dislikeStatus: PropTypes.bool,
      }).isRequired
    ).isRequired,

  };

  componentDidMount() {
    const { articleSlug, getComments } = this.props;
    getComments(articleSlug);
  }


  render() {
    const {
      comments, commentsCount, articleSlug, deleteComment, username
    } = this.props;

    const commentList = comments.reverse().map(comment => (
      <CommentBox
        key={comment.id}
        {...comment}
        deleteComment={deleteComment}
        articleSlug={articleSlug}
        authenticatedUsername={username}

      />
    ));


    return (
      <div className="commentForm">
        <span className="comment_section_stats">
          {commentsCount}
          {' '}
          comments
        </span>
        <CommentForm articleSlug={articleSlug} />
        <div className="comments_list">
          {' '}
          {commentList}
        </div>
      </div>
    );
  }
}

// map state to props
export const mapStateToProps = (state) => {
  const {
    commentReducer: { comments, commentsCount },
    loginReducer: { user: { username } },
  } = state;
  return { comments, commentsCount, username, };
};

// map dispatch to props
export const mapDispatchToProps = dispatch => ({
  getComments(articleSlug) {
    dispatch(readComments(articleSlug));
  },
  deleteComment(articleSlug, commentId) {
    dispatch(removeComment(articleSlug, commentId));
  }
});

// export default CommentList;
export default connect(mapStateToProps, mapDispatchToProps)(Comment);
