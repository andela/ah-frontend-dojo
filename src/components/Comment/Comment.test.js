import React from 'react';
import { shallow } from 'enzyme';
import { Comment, mapStateToProps, mapDispatchToProps } from 'components/Comment/';
import commentTestData from 'store/reducers/commentReducer/__mocks__';


describe('Comment Box', () => {
  let wrapper;
  const getCommentsFn = jest.fn();
  const deleteCommentFn = jest.fn();
  const props = {
    getComments: getCommentsFn,
    deleteComment: deleteCommentFn,
    comments: commentTestData,
    articleSlug: 'javascript-code',
    commentsCount: commentTestData.length,
    username: 'zack'
  };

  it('should render without crushing', () => {
    wrapper = shallow(<Comment
      {...props}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should map state to props', () => {
    const mockedState = {
      loginReducer: {
        user: {
          username: 'arthur'
        }
      },
      commentReducer: {
        comments: commentTestData,
        commentsCount: commentTestData.length,
      }
    };

    const state = mapStateToProps(mockedState);

    expect(state.username).toEqual('arthur');
    expect(state.commentsCount).toBe(2);
  });

  it('should map dispatch to props', () => {
    mapDispatchToProps(getCommentsFn).getComments();
    expect(getCommentsFn).toHaveBeenCalled();

    mapDispatchToProps(deleteCommentFn).deleteComment();
    expect(deleteCommentFn).toHaveBeenCalled();
  });
});
