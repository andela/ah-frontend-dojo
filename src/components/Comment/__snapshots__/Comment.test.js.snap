// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Box should render without crushing 1`] = `
<div
  className="commentForm"
>
  <span
    className="comment_section_stats"
  >
    2
     
    comments
  </span>
  <Connect(CommentForm)
    articleSlug="javascript-code"
  />
  <div
    className="comments_list"
  >
     
    <CommentBox
      articleSlug="javascript-code"
      authenticatedUsername="zack"
      author={
        Object {
          "bio": "This is arthur biodata information",
          "email": "email2@ahhaven.com",
          "firstname": "Firstnamex",
          "image": "https://firebasestorage.googleapis.com/v0/b/ah-frontend-dojo.appspot.com/o/images%2F61030656_367914100741361_8198565189029724160_n.jpg?alt=media&token=a2c313ca-3c98-44f6-b89e-bba30409c49f",
          "lastname": "Lastnamex",
          "username": "adojo",
        }
      }
      body="my second comment"
      created_at="2019-06-14T07:43:08.288832Z"
      deleteComment={[MockFunction]}
      dislikeStatus={false}
      end_index={null}
      highlighted_text={null}
      id={2}
      key="2"
      likeStatus={false}
      start_index={null}
      updated_at="2019-06-14T07:43:08.288889Z"
    />
    <CommentBox
      articleSlug="javascript-code"
      authenticatedUsername="zack"
      author={
        Object {
          "bio": "This is arthur biodata information",
          "email": "email1@ahhaven.com",
          "firstname": "Firstname",
          "image": "https://firebasestorage.googleapis.com/v0/b/ah-frontend-dojo.appspot.com/o/images%2F61030656_367914100741361_8198565189029724160_n.jpg?alt=media&token=a2c313ca-3c98-44f6-b89e-bba30409c49f",
          "lastname": "Lastname",
          "username": "adojo",
        }
      }
      body="my first comment"
      created_at="2019-06-14T05:58:57.206787Z"
      deleteComment={[MockFunction]}
      dislikeStatus={false}
      end_index={null}
      highlighted_text={null}
      id={1}
      key="1"
      likeStatus={false}
      start_index={null}
      updated_at="2019-06-14T05:58:57.206825Z"
    />
  </div>
</div>
`;
